pkg/service/did/ion.go

```diff
@@ -658,3 +728,82 @@ func (h *ionHandler) storeUpdateStates(ctx context.Context, tx storage.Tx, id st
 	}
 	return nil
 }
+
+func (h *ionHandler) DeactivateDID(ctx context.Context, request DeactivateIONDIDRequest) (*DeactivateIONDIDResponse, error) {
+	storedDID := new(ionStoredDID)
+	if err := h.storage.GetDID(ctx, request.DID.String(), storedDID); err != nil {
+		return nil, errors.Wrap(err, "getting ion did from storage")
+	}
+
+	if storedDID.Deactivated {
+		return &DeactivateIONDIDResponse{Result: storedDID.DeactivationResponse}, nil
+	}
+
+	privateUpdateJWK, err := h.readUpdatePrivateKey(ctx, request.DID.String())
+	if err != nil {
+		return nil, err
+	}
+	privateRecoveryJWK, err := h.readRecoveryPrivateKey(ctx, request.DID.String())
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := ion.NewBTCSignerVerifier(*privateUpdateJWK)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating signer")
+	}
+
+	suffix, err := request.DID.Suffix()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting suffix")
+	}
+	deactivateRequest, err := ion.NewDeactivateRequest(suffix, privateRecoveryJWK.ToPublicKeyJWK(), *signer)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating deactivate request")
+	}
+
+	if _, err := h.resolver.Anchor(ctx, deactivateRequest); err != nil {
+		return nil, errors.Wrap(err, "anchoring deactivate operation")
+	}
+
+	const deactivateRequestNamespace = "deactivate-request"
+	deactivateRequestKey := request.DID.String()
+	watchKeys := []storage.WatchKey{
+		{
+			Namespace: deactivateRequestNamespace,
+			Key:       deactivateRequestKey,
+		},
+	}
+	deactivatedDID, err := h.storage.db.Execute(ctx, func(ctx context.Context, tx storage.Tx) (any, error) {
+		if storedDID.Deactivated {
+			return storedDID, nil
+		}
+		storedDID.Operations = append(storedDID.Operations, deactivateRequest)
```

Comment:
this seems off - shouldn't the result be the processed deactivate request?

Reply:
disregard

Reply:
I think we're eventually going to need to communicate to devs that some new state hasn't been published. That said, I agree that it's too complex for now.

I made my best to ensure that all the properties were correct.